Library/Formula/elm.rb

```diff
@@ -0,0 +1,98 @@
+require "language/haskell"
+
+class Elm < Formula
+  include Language::Haskell::Cabal
+
+  desc "Functional programming language for building browser-based GUIs"
+  homepage "http://elm-lang.org"
+
+  stable do
+    url "https://github.com/elm-lang/elm-compiler/archive/0.16.tar.gz"
+    sha256 "ea4ff37ec6a1bfb8876e7a9b2aa0755df9ac92f5e5c8bfcc611b1886fb06bb13"
+
+    resource "elm-package" do
+      url "https://github.com/elm-lang/elm-package/archive/0.16.tar.gz"
+      sha256 "1cac7d27415a4d36d7b1c7260953e0c7b006e7cbb24d5bdb3b0d440d375a8bf5"
+    end
+
+    resource "elm-make" do
+      url "https://github.com/elm-lang/elm-make/archive/0.16.tar.gz"
+      sha256 "ed2eb38ee3d41307751b9df4fd464987c7cdd96413a907b800923af8a25a8c15"
+    end
+
+    resource "elm-repl" do
+      url "https://github.com/elm-lang/elm-repl/archive/0.16.tar.gz"
+      sha256 "ea89a6cccd546e26c6af12bd35b886bfa666888323596c60168013acf67fe2e4"
+    end
+
+    resource "elm-reactor" do
+      url "https://github.com/elm-lang/elm-reactor/archive/0.16.tar.gz"
+      sha256 "fd105180b92364c3ba8df58bfa88306adb8a21074786267ab1eaa634777ded7f"
+    end
+  end
+
+  head do
+    url "https://github.com/elm-lang/elm-compiler.git"
+
+    resource "elm-package" do
+      url "https://github.com/elm-lang/elm-package.git"
+    end
+
+    resource "elm-make" do
+      url "https://github.com/elm-lang/elm-make.git"
+    end
+
+    resource "elm-repl" do
+      url "https://github.com/elm-lang/elm-repl.git"
+    end
+
+    resource "elm-reactor" do
+      url "https://github.com/elm-lang/elm-reactor.git"
+    end
+  end
+
+  depends_on "ghc" => :build
+  depends_on "cabal-install" => :build
+
+  setup_ghc_compilers
+
+  def install
+    compiler_path = buildpath/"elm-compiler"
+    mktemp do
+      mv Dir["#{buildpath}/*"], pwd
+      mv pwd, compiler_path
+    end
```

Comment:
Ultimately, we want to have the following directory structure inside the buildpath:
```
elm-compiler/...
elm-package/...
elm-make/...
elm-repl/...
elm-reactor/...
```
Each of these subdirectories contains one Cabal package.  In the next step of the build process, we add each of these packages to a single Cabal sandbox.  This allows Cabal to find the correct version of these packages, and to successfully resolve dependencies.

Being able to ask Homebrew to unpack in a subdirectory would be nice, but it’s not required, as the formula is correct, if slightly inelegant.